In [1]:
import pandas as pd
import numpy as np
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV, train_test_split
import os

In [2]:
filename = "Loan.csv"
df = pd.read_csv(filename)
print(df.shape)
df.head()

(16198, 36)


loan_amnt  installment  int_rate  emp_length  annual_inc  delinq_2yrs  \
0  -0.677420    -0.647187 -1.343269   -1.389984    0.632875    -0.296965   
1   0.489425     0.802106  0.176176    1.424483   -0.589384    -0.296965   
2   0.492857     0.754797 -0.240926   -1.108537   -0.112405    -0.296965   
3  -1.020610    -0.980425 -0.005290   -1.389984   -0.373253    -0.296965   
4  -1.281435    -1.272871  0.834332   -1.389984   -0.112405    -0.296965   

   earliest_cr_line  open_acc  inq_last_6mths       dti  ...  x2_medical  \
0          0.481628  0.855178        0.120850 -0.552658  ...   -0.136136   
1         -0.064506 -1.416934       -0.808952  0.279006  ...   -0.136136   
2          0.392254 -1.416934       -0.808952  1.973746  ...   -0.136136   
3          0.681987 -1.189722       -0.808952 -0.983448  ...   -0.136136   
4          0.972086 -0.053667        1.050653 -1.023834  ...   -0.136136   

   x2_moving  x2_other  x2_renewable_energy  x2_small_business  x2_vacation  \
0  -0.124993 -0.334951             -0.05278          -0.218828    -0.100154   
1  -0.124993 -0.334951             -0.05278          -0.218828    -0.100154   
2  -0.124993 -0.334951             -0.05278          -0.218828    -0.100154   
3  -0.124993 -0.334951             -0.05278          -0.218828    -0.100154   
4  -0.124993 -0.334951             -0.05278          -0.218828    -0.100154   

   x2_wedding  x3_Source Verified  x3_Verified  loan_status  
0    6.340257            1.727971    -0.674833          0.0  
1   -0.157722           -0.578714    -0.674833          0.0  
2   -0.157722            1.727971    -0.674833          0.0  
3   -0.157722           -0.578714    -0.674833          0.0  
4   -0.157722           -0.578714    -0.674833          0.0  

[5 rows x 36 columns]

In [3]:

#df.drop(0, inplace = True)
columns = df.columns
features = columns.drop('loan_status')
target = ['loan_status']
x = df[features]
y=df[target]

In [4]:
print(features.size)
y.head()

35


loan_status
0          0.0
1          0.0
2          0.0
3          0.0
4          0.0

In [5]:
x = x.astype(float)
y = y.astype(float)
#train_x, test_x, train_y, test_y = train_test_split(x,y,test_size = 0.2, random_state=0)

train = pd.read_csv("../src/resources/datasets/Loan/Loan.train5.csv")
test = pd.read_csv("../src/resources/datasets/Loan/Loan.test5.csv")
train_x = train[features]
train_y = train[target]
test_x = train[features]
test_y = train[target]


In [6]:
train_y.head()

loan_status
0          0.0
1          0.0
2          0.0
3          0.0
4          0.0

In [7]:
n_estimators = 10
max_depth = 3
myclassifier = RandomForestClassifier(random_state=42, criterion = 'gini', n_estimators = n_estimators, max_features = 'sqrt', max_depth =max_depth)
params = {}
n_fold = 5

gscv_rf = GridSearchCV(myclassifier, param_grid = params, scoring='recall', cv=n_fold, refit=True)

gscv_rf.fit(train_x, train_y.loan_status)


GridSearchCV(cv=5, error_score=nan,
             estimator=RandomForestClassifier(bootstrap=True, ccp_alpha=0.0,
                                              class_weight=None,
                                              criterion='gini', max_depth=3,
                                              max_features='sqrt',
                                              max_leaf_nodes=None,
                                              max_samples=None,
                                              min_impurity_decrease=0.0,
                                              min_impurity_split=None,
                                              min_samples_leaf=1,
                                              min_samples_split=2,
                                              min_weight_fraction_leaf=0.0,
                                              n_estimators=10, n_jobs=None,
                                              oob_score=False, random_state=42,
                                      

In [8]:
rf = gscv_rf.best_estimator_
trees = rf.estimators_
for tree in trees:
    print(tree.tree_.node_count)

15
15
15
15
15
15
15
15
15
15


In [9]:
print(trees[1].tree_.feature[3])
print(trees[1].tree_.value)
print(trees[1].tree_.value[0])
print(trees[1].tree_.impurity[0])
print(trees[1].tree_.value[14])



-2
[[[4.216e+03 4.266e+03]]

 [[2.072e+03 1.765e+03]]

 [[3.430e+02 4.650e+02]]

 [[9.000e+00 3.700e+01]]

 [[3.340e+02 4.280e+02]]

 [[1.729e+03 1.300e+03]]

 [[1.252e+03 8.450e+02]]

 [[4.770e+02 4.550e+02]]

 [[2.144e+03 2.501e+03]]

 [[2.091e+03 2.374e+03]]

 [[9.050e+02 1.112e+03]]

 [[1.186e+03 1.262e+03]]

 [[5.300e+01 1.270e+02]]

 [[5.200e+01 1.170e+02]]

 [[1.000e+00 1.000e+01]]]
[[4216. 4266.]]
0.4999826254535382
[[ 1. 10.]]


DATASET_NAME: Breast-Cancer-Wisconsin.train1.csv
ENSEMBLE: RF
NB_TREES: 10
NB_FEATURES: 9
NB_CLASSES: 2
MAX_TREE_DEPTH: 3
Format: node / node type (LN - leave node, IN - internal node) left child / right child / feature / threshold / node_depth / majority class (starts with index 0)


In [10]:
firstTree = trees[0].tree_
n_nodes = firstTree.node_count
children_left = firstTree.children_left
children_right = firstTree.children_right
feature = firstTree.feature
threshold = firstTree.threshold

node_depth = np.zeros(shape=n_nodes, dtype=np.int64)
is_leaves = np.zeros(shape=n_nodes, dtype=bool)
stack = [(0, 0)]  # start with the root node id (0) and its depth (0)
while len(stack) > 0:
    # `pop` ensures each node is only visited once
    node_id, depth = stack.pop()
    node_depth[node_id] = depth

    # If the left and right child of a node is not the same we have a split
    # node
    is_split_node = children_left[node_id] != children_right[node_id]
    # If a split node, append left and right children and depth to `stack`
    # so we can loop through them
    if is_split_node:
        stack.append((children_left[node_id], depth + 1))
        stack.append((children_right[node_id], depth + 1))
    else:
        is_leaves[node_id] = True

print(
    "The binary tree structure has {n} nodes and has "
    "the following tree structure:\n".format(n=n_nodes)
)
for i in range(n_nodes):
    if is_leaves[i]:
        print(firstTree.value[i])
        print(firstTree.impurity[i])
        
    else:
        print("")
        

The binary tree structure has 15 nodes and has the following tree structure:




[[877. 393.]]
0.4273804947609895
[[838. 675.]]
0.4941968024204416

[[1603. 1217.]]
0.49063201046225036
[[ 60. 110.]]
0.45674740484429066


[[439. 958.]]
0.43099001900483136
[[ 21. 103.]]
0.28134755463059313

[[478. 645.]]
0.48894284407067
[[ 9. 56.]]
0.23857988165680477


In [11]:
len(trees)

10

In [34]:
scriptDir= os.path.abspath('')
parent = os.path.abspath(os.path.join(scriptDir, os.pardir))
relativePath = "src/resources/forests/Loan/" +filename[0:len(filename)-3] + "RF" + str(n_fold) + ".txt"
absolutePath = os.path.join(parent, relativePath)
print(absolutePath)

/home/fabiano/Scaricati/BA-Trees-master/src/resources/forests/Loan/Loan.RF5.txt


In [35]:
f = open(str(absolutePath), 'w')
f.write("DATASET_NAME: " + filename+"\n")
f.write("ENSEMBLE: RF\n")
f.write("NB_TREES: " + str(n_estimators)+"\n")
f.write("NB_FEATURES: " + str(features.size)+"\n")
f.write("NB_CLASSES: 2\n")
f.write("MAX_TREE_DEPTH: " + str(max_depth)+"\n")
f.write("Format: node / node type (LN - leave node, IN - internal node) left child / right child / feature / threshold / node_depth / majority class (starts with index 0)\n")

for j in range(0, len(trees)):
    f.write("\n[TREE " + str(j)+"]\n")
    tree = trees[j].tree_
    n_nodes = tree.node_count
    f.write("NB_NODES: " + str(n_nodes)+"\n")

    children_left = tree.children_left
    children_right = tree.children_right
    feature = tree.feature
    threshold = tree.threshold
    value = tree.value

    node_depth = np.zeros(shape=n_nodes, dtype=np.int64)
    is_leaves = np.zeros(shape=n_nodes, dtype=bool)
    stack = [(0, 0)]  # start with the root node id (0) and its depth (0)
    while len(stack) > 0:
        # `pop` ensures each node is only visited once
        node_id, depth = stack.pop()
        node_depth[node_id] = depth

        # If the left and right child of a node is not the same we have a split
        # node
        is_split_node = children_left[node_id] != children_right[node_id]
        # If a split node, append left and right children and depth to `stack`
        # so we can loop through them
        if is_split_node:
            stack.append((children_left[node_id], depth + 1))
            stack.append((children_right[node_id], depth + 1))
        else:
            is_leaves[node_id] = True

    for i in range(n_nodes):
        f.write(str(i) + " ")
        if is_leaves[i]:
            f.write("LN ")
            f.write(str(children_left[i])+ " " + str(children_right[i]) + " " + str(feature[i]) + " " + str(threshold[i]) + " " + str(node_depth[i]) + " ")
            if(value[i][0][0]>=value[i][0][1]):
                f.write("0")
            else:
                f.write("1")
            
        else:
            f.write("IN ")
            f.write(str(children_left[i])+ " " + str(children_right[i]) + " " + str(feature[i]) + " " + str(threshold[i]) + " " + str(node_depth[i]) + " ")
            f.write("-1")
        f.write("\n")
f.write("\n")
f.flush()
f.close()
    
    



[TREE 0]
NB_NODES: 13
0 IN 1 6 2 2.5 0 -1
1 IN 2 5 0 8.0 1 -1
2 IN 3 4 4 9.0 2 -1
3 LN -1 -1 -1 -1 3 0
4 LN -1 -1 -1 -1 3 0
5 LN -1 -1 -1 -1 2 1
6 IN 7 10 6 2.5 1 -1
7 IN 8 9 7 4.0 2 -1
8 LN -1 -1 -1 -1 3 0
9 LN -1 -1 -1 -1 3 1
10 IN 11 12 4 2.5 2 -1
11 LN -1 -1 -1 -1 3 1
12 LN -1 -1 -1 -1 3 1